### DISCUSSION TITLE
which liburing library version should I use for kernel  5.10?

### DISCUSSION THREAD

**Discussion by osevan** [axboe/liburing#585] [Q&A] [answered]
Which version is latest stable one for kernel 5.10?

Thanks and

Best regards

**Comment by axboe** [comment, marked as answer]
Just use the latest one, liburing and the kernel aren't tied to each other, it's just that some features are only available on newer kernels. But you can use the 2.1 release with 5.10 just fine, as well as the upcoming 2.2.

**Reply by victorstewart** [reply]
but it'll fail to build if new opcodes were added to liburing to support newer kernel operations, because the enum values aren't in the older kernel's header file. i've encountered that a few times.

**Reply by osevan** [reply]
> Just use the latest one, liburing and the kernel aren't tied to each other, it's just that some features are only available on newer kernels. But you can use the 2.1 release with 5.10 just fine, as well as the upcoming 2.2.

New version are more performing better than normal versions? 

This is important too for me.


Thanks and

Best regards

**Reply by axboe** [reply]
The current master branch, which will become 2.2, does have a few optimizations over the 2.1 release. So yes, it should be better. As usual, also fixes bugs found in 2.1.

**Comment by axboe** [comment]
> but it'll fail to build if new opcodes were added to liburing to support newer kernel operations, because the enum values aren't in the older kernel's header file. i've encountered that a few times.

How so? liburing doesn't depend on any kernel io_uring headers, it carries its own for that very reason.

**Reply by victorstewart** [reply]
hmmm. downloading the source from github and building it with the standard sequence of steps has always produced those errors for me. like "can't find definition of IORING_OP_XYZ" which were for-next type operations.

i'll try it tomorrow and report back.

**Reply by axboe** [reply]
That sounds very odd, and honestly nobody has ever reported anything like that or it would have been fixed immediately. liburing is completely standalone, and has no dependencies on the kernel (or the io_uring header from there), very deliberately so.

**Reply by axboe** [reply]
Hence if there is an issue you are seeing, please do report it so it can get fixed. Maybe it's specific to your build environment?

**Reply by victorstewart** [reply]
it's been like this for me for years now ever since liburing was born, and i just chalked it up to some compatibility thing. that's why i chimed in here.

but i'll definitely report back tomorrow with details. i'm on 5.17 now so as long as l there are new op codes in 5.18 or for-next we'll see it.

**Reply by axboe** [reply]
You have to report things like that! Not reporting something means it never gets fixed. That's a black mark.

Both 5.18 and 5.19-next has new opcodes over 5.17, so if that's the issue for you, it should definitely trigger.

**Reply by victorstewart** [reply]
i thought it was a feature not a bug 😂

**Reply by axboe** [reply]
I appreciate the glass half full mentality ;-)

**Reply by victorstewart** [reply]
okay correction it's not when it builds, but when you run the application with the liburing header, that if you're using a version with future opcodes that the definitions are missing.

```
/xxx/libraries/include/liburing/liburing.h:307:7: error: no member named 'addr3' in 'io_uring_sqe'
        sqe->addr3 = 0;
        ~~~  ^
/xxx/libraries/include/liburing/liburing.h:529:38: error: use of undeclared identifier 'IORING_ASYNC_CANCEL_FD'; did you mean 'IORING_OP_ASYNC_CANCEL'?
        sqe->cancel_flags = (__u32) flags | IORING_ASYNC_CANCEL_FD;
                                            ^~~~~~~~~~~~~~~~~~~~~~
                                            IORING_OP_ASYNC_CANCEL
/usr/include/linux/io_uring.h:117:2: note: 'IORING_OP_ASYNC_CANCEL' declared here
        IORING_OP_ASYNC_CANCEL,
        ^
In file included from /root/nametag/application/application.server.cpp:5:
In file included from /xxx/networking/includes.h:15:
/xxx/libraries/include/liburing/liburing.h:746:19: error: use of undeclared identifier 'IORING_OP_MSG_RING'
        io_uring_prep_rw(IORING_OP_MSG_RING, sqe, fd, NULL, len, data);
                         ^
/xxx/libraries/include/liburing/liburing.h:756:19: error: use of undeclared identifier 'IORING_OP_GETXATTR'; did you mean 'IORING_OP_STATX'?
        io_uring_prep_rw(IORING_OP_GETXATTR, sqe, 0, name, len,
                         ^~~~~~~~~~~~~~~~~~
                         IORING_OP_STATX
/usr/include/linux/io_uring.h:124:2: note: 'IORING_OP_STATX' declared here
        IORING_OP_STATX,
        ^
```

**Reply by victorstewart** [reply]
so until now i've just been using git checkout with a commit tag that includes all the features i use/need

**Reply by victorstewart** [reply]
haha given the self contained point you made, now i'm guessing there's a header with all these definitions that i haven't been including?

**Reply by axboe** [reply]
This is not making sense to me. This is clearly at compile time, not runtime. Are you using a liburing.h with an io_uring.h from an older kernel somehow? This sounds like a build issue, something specific on your end? You have a non-coherent set of headers, are you including io_uring.h from your app somehow before liburing.h? Just guessing here.

I had to look to see what system header I have it installed, it's a lot older than the liburing once I'm building. I built all versions since 0.7 and up and I don't see anything like this.

**Reply by axboe** [reply]
> haha given the self contained point you made, now i'm guessing there's a header with all these definitions that i haven't been including?

You don't need to include anything outside of liburing.h, in fact my strong suspicion above is that you are including some unrelated io_uring.h that came from another package?

**Reply by axboe** [reply]
How did you install liburing? Package or self-installed? liburing puts the header in install-path/include/liburing.h and the other headers from liburing end up in liburing/.

**Reply by victorstewart** [reply]
right i meant when you compile the application. ahh you solved it! from way back when i had this line `#include <linux/io_uring.h>` included 🙄

**Reply by axboe** [reply]
Mystery solved! Back to the original question, my original comment still stands. Use the latest version, there are no kernel dependencies. EOD ;-)
